# --- ROCK-COPYRIGHT-NOTE-BEGIN ---
# 
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# Please add additional copyright information _after_ the line containing
# the ROCK-COPYRIGHT-NOTE-END tag. Otherwise it might get removed by
# the ./scripts/Create-CopyPatch script. Do not edit this copyright text!
# 
# ROCK Linux: rock-src/package/x11/xfree86/radeon-iBook.patch
# ROCK Linux is Copyright (C) 1998 - 2004 Clifford Wolf
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# 
# --- ROCK-COPYRIGHT-NOTE-END ---


On iBook the PLL values for the LCD port are not caluculated correctly - since
I do not have the databook I applied the same hack as present in the BenH
kernel radeon fb and added a new default value case ...

  - Rene Rebe <rene@rocklinux.org>

diff -ur xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c xc-working/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c
--- xc/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c	2004-02-21 18:25:41.000000000 +0100
+++ xc-working/programs/Xserver/hw/xfree86/drivers/ati/radeon_driver.c	2004-02-21 18:27:10.000000000 +0100
@@ -1808,6 +1808,11 @@
 		    pll->reference_div  = 12;
 		    pll->xclk           = 23000;
 		    break;
+		case PCI_CHIP_RADEON_LW: /* Guessed based on iBook Open Firmware data -ReneR */
+		    pll->reference_freq = 2700;
+		    pll->reference_div  = 12;
+		    pll->xclk           = 36000;
+		    break;
 		default:
 		    pll->reference_freq = 2700;
 		    pll->reference_div  = 67;
@@ -6064,7 +6069,7 @@
     save->disp_merge_cntl &= ~RADEON_DISP_RGB_OFFSET_EN;
 
 #if X_BYTE_ORDER == X_BIG_ENDIAN
-    /* Alhought we current onlu use aperture 0, also setting aperture 1 should not harm -ReneR */
+    /* Alhought we currently only use aperture 0, setting aperture 1 should not harm. -ReneR */
     switch (pScrn->bitsPerPixel) {
     case 16:
 	save->surface_cntl |= RADEON_NONSURF_AP0_SWP_16BPP;
@@ -6507,8 +6512,17 @@
 	       save->feedback_div,
 	       save->post_div));
 
+    ErrorF ("ref_div: %d, bitvalue: %d, feedback_div: %d\n",
+            pll->reference_div, post_div->bitvalue, save->feedback_div);
+
     save->ppll_ref_div   = pll->reference_div;
     save->ppll_div_3     = (save->feedback_div | (post_div->bitvalue << 16));
+#if defined(__powerpc__) /* on iBooks the LCD pannel needs tweaked PLL timings -ReneR */
+    ErrorF ("pll_div_3: %d, ", save->ppll_div_3);
+    /* save->ppll_div_3     *= 6; */ /* just a test hack */
+    save->ppll_div_3     = 0x000600ad; /* -ReneR */
+    ErrorF ("is now: %d, should be: %d\n", save->ppll_div_3, 0x000600ad);
+#endif
     save->htotal_cntl    = 0;
 }
 
